### PR TITLE
Handle `set` edge cases for Dimension/DimTuple args

### LINKED ISSUE #478: Inconsistency of `set`
When the 2nd argument of `set` is a `Dimension`, `set` is inconsistent about whether it updates the name of the dimension or not:

```julia
julia> lu = LookupArrays.Sampled(1:10);

julia> set(Dim{:foo}(lu), :bar)  # (✓) changes name (✓) keeps lookup
Dim{:bar} Sampled{Int64} AutoOrder AutoSpan AutoSampling
wrapping: 1:10

julia> set(Dim{:foo}(lu), Dim{:bar})  # (✓) changes name (✓) keeps lookup
Dim{:bar} Sampled{Int64} AutoOrder AutoSpan AutoSampling
wrapping: 1:10

julia> set(Dim{:foo}(lu), Dim{:bar}(2:11))  # (✓) changes name (✓) changes lookup
Dim{:bar} Sampled{Int64} ForwardOrdered Regular AutoSampling
wrapping: 2:11

julia> set(Dim{:foo}(lu), Dim{:bar}(LookupArrays.Sampled(2:11)))  # (✗) changes name (✓) changes lookup
Dim{:foo} Sampled{Int64} ForwardOrdered Regular AutoSampling
wrapping: 2:11

julia> set(Dim{:foo}(lu), Dim{:bar}(lu))   # (✗) changes name (✓) lookup unchanged
Dim{:foo} Sampled{Int64} ForwardOrdered Regular AutoSampling
wrapping: 1:10
```

As a bonus, the following cases just error instead of updating the dimension name:
```julia
julia> set(Dim{:foo}(), :bar)
ERROR: ArgumentError: Can not set any fields of Colon to Colon
Stacktrace:
 [1] _cantseterror(a::Function, b::Function)
   @ DimensionalData.Dimensions.LookupArrays ~/projects/DimensionalData.jl/src/LookupArrays/set.jl:99
 [2] _set(A::Function, x::Function)
   @ DimensionalData.Dimensions.LookupArrays ~/projects/DimensionalData.jl/src/LookupArrays/set.jl:96
 [3] _set(dim::Dim{:foo, Colon}, newdim::Dim{:bar, Colon})
   @ DimensionalData.Dimensions ~/projects/DimensionalData.jl/src/Dimensions/set.jl:24
 [4] _set(dim::Dim{:foo, Colon}, key::Symbol)
   @ DimensionalData.Dimensions ~/projects/DimensionalData.jl/src/Dimensions/set.jl:27
 [5] set(dim::Dim{:foo, Colon}, x::Symbol)
   @ DimensionalData.Dimensions ~/projects/DimensionalData.jl/src/Dimensions/set.jl:3
 [6] top-level scope
   @ REPL[506]:1

julia> set(Dim{:foo}(1:10), :bar)
ERROR: ArgumentError: Can not set any fields of UnitRange{Int64} to Colon
Stacktrace:
 [1] _cantseterror(a::UnitRange{Int64}, b::Function)
   @ DimensionalData.Dimensions.LookupArrays ~/projects/DimensionalData.jl/src/LookupArrays/set.jl:99
 [2] _set(A::UnitRange{Int64}, x::Function)
   @ DimensionalData.Dimensions.LookupArrays ~/projects/DimensionalData.jl/src/LookupArrays/set.jl:96
 [3] _set(dim::Dim{:foo, UnitRange{Int64}}, newdim::Dim{:bar, Colon})
   @ DimensionalData.Dimensions ~/projects/DimensionalData.jl/src/Dimensions/set.jl:24
 [4] _set(dim::Dim{:foo, UnitRange{Int64}}, key::Symbol)
   @ DimensionalData.Dimensions ~/projects/DimensionalData.jl/src/Dimensions/set.jl:27
 [5] set(dim::Dim{:foo, UnitRange{Int64}}, x::Symbol)
   @ DimensionalData.Dimensions ~/projects/DimensionalData.jl/src/Dimensions/set.jl:3
 [6] top-level scope
   @ REPL[507]:1
```

### PR DESCRIPTION
**Pull request by sethaxen**
Fixes #478

### CODE REVIEW

**Comment by rafaqz** [comment]
Will review in a few days im travelling

File: test/set.jl [hunk 1]
```diff
@@ -112,6 +112,18 @@ end
         @test order(set(uda, X=ReverseOrdered())) == (ReverseOrdered(), Unordered())
     end
 
+    @testset "issue #478" begin
```
Comment (rafaqz) [inline test/set.jl:115]: Maybe add a more descriptive testset name for future readers

**Review by rafaqz — APPROVED** [review]
Thanks, looks good

File: test/set.jl (hunk 1 above)
Comment (sethaxen) [reply test/set.jl:115]:
```suggestion
    # issue #478
    @testset "tuple dims and/or Symbol/Dim{Colon}/Colon replacement" begin
```

**Pull request merged**